pytorch/train.py

```diff
@@ -0,0 +1,186 @@
+# *****************************************************************************
+#  Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
+# 
+#  Redistribution and use in source and binary forms, with or without
+#  modification, are permitted provided that the following conditions are met:
+#      * Redistributions of source code must retain the above copyright
+#        notice, this list of conditions and the following disclaimer.
+#      * Redistributions in binary form must reproduce the above copyright
+#        notice, this list of conditions and the following disclaimer in the
+#        documentation and/or other materials provided with the distribution.
+#      * Neither the name of the NVIDIA CORPORATION nor the
+#        names of its contributors may be used to endorse or promote products
+#        derived from this software without specific prior written permission.
+# 
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+#  ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+#  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+#  DISCLAIMED. IN NO EVENT SHALL NVIDIA CORPORATION BE LIABLE FOR ANY
+#  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+#  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+#  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+#  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+#  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+# 
+# *****************************************************************************
+import argparse
+import json
+import os
+import time
+import torch
+
+#=====START: ADDED FOR DISTRIBUTED======
+from distributed import init_distributed, apply_gradient_allreduce, reduce_tensor
+from torch.utils.data.distributed import DistributedSampler
+#=====END:   ADDED FOR DISTRIBUTED======
+
+from torch.utils.data import DataLoader
+from wavenet import WaveNet
+from mel2samp_onehot import Mel2SampOnehot
+from utils import to_gpu
+
+class CrossEntropyLoss(torch.nn.Module):
+    def __init__(self):
+        super(CrossEntropyLoss, self).__init__()
+        self.num_classes = wavenet_config["n_out_channels"]
+
+    def forward(self, inputs, targets):
+        targets = targets.view(-1)
```

Comment:
Add a couple comments as towards these random tensor operations.

Reply:
Do you think these operations are random or are you just expressing your dissatisfaction in not understanding the code? If you mean the second, let us know what is unclear and we'll make it clearer.

Reply:
Uhhh. I hope that above comment does not come off as ingenuine. 

I understand that motivation for the above operations because I read the entire above code and replicated WaveNet on my own.

From a locality standpoint of view, the method documentation is not self-sufficient. It's typically bad code style to have functions that do not make sense on their own. 

I understand the code works. I'm happy to drop the issue and move on. It's just a code style thing.

Reply:
I've added some comments to make it less opaque.  I'm leaving the names `inputs` and `targets` for now as that's what they use in CrossEntropyLoss